refactor: Make get notes return all notes at beginning of array #4991

sklppy88 opened the pull request:
Resolves #4991

Commit:
Fix

Commit:
fix

Commit:
fix

Commit:
asdf

LHerskind reviewed (commented):
Can you create a test that explicitly make this return a different array that what the oracle produces 👍

Commit:
add test

Commit:
add test

sklppy88 committed:
Merge branch 'master' into ek/refactor/get-notes-returns-non-sparse-array

Commit:
fix

Commit:
formatting fix

MirandaWood committed:
feat: truncate SHA hashes inside circuits (#5160)

Will close #2019

This PR converts SHA hashing inside noir circuits from outputting 2
128-bit fields to outputting 1 248-bit field. To fit inside the field,
we truncate one byte.

---
### Noir Changes

The constant `NUM_FIELDS_PER_SHA256` is now 1, so any hardcoded test
values and function returns have been changed to use an array of one.
I've kept it as an array rather than a single `Fr` to minimise changes
across the repo and ensure if we want to revert `NUM_FIELDS_PER_SHA256`
in future, it won't be so painful. However, we can also just use a
single `Fr` if that's preferable.

`TX_EFFECTS_HASH_LOG_FIELDS`

Methods:

- `field_from_bytes_32_trunc`: Converts a 32 byte array to a 31 byte
field element (useful for comparisons with new `sha256_to_field`), tests
in `types/src/utils/field.nr`.
- `sha256_to_field`: Uses the same method as the previous version to
convert the sha result (BE) bytes array to field, but leaves out the
final byte.
- `accumulate_sha256`: Used almost exclusively for enc/unenc logs
hashing - takes in 2 31 byte field elements, assumed to be outputs of a
previous sha hash, pads to 32 bytes and hashes them with
`sha256_to_field` as a 64 byte array. Note that as before, other
circuits that use sha (like tx effects hash and messages hash) do not
use this method and instead create a flat byte array, then call
`sha256_to_field`.

---
### L1 Contract Changes

To match the Noir method, the `sha256ToField` function now truncates a
byte and prepends a blank byte. Not prepending the blank byte means
changing many struct fields from `bytes32` to `bytes31`. This (IIRC) is
the same gas cost and creates more awkward encoding, so I kept the
length with a blank byte. This also changes the slither file, as I
removed some of the old encoding which flagged with new encoding...
which also flags.

~Only the 'leaves' used in computing the `txsHash` in `TxsDecoder` and
logs hashes have been changed to 31 bytes to match the Noir SHA
accumulation (since we are repeating hashes of hashes).~

~The TS code (see below) does pack the Header struct with 31 bytes per
SHA, so we must shift the decoding in HeaderLib` by 3 bytes.~

As of 21.3, there have been a lot of changes in master to the way the
txs effect hash (formerly calldata hash/txs hash) is calculated. Plus,
now we actually recalculate the in/outHash (i.e. the root of the sha
tree of messages) in the contract, so I have reverted to using 32 bytes
everywhere with a prepended blank byte.

---
### TS Changes

All `.hash()` methods which are also computed in the circuit have been
changed to match the Noir code. In most places this just means
truncating a byte with `.subarray(0, 31)` on the buffer.
~The `ContentCommitment` serialise/deserialise methods have been
modified, as keeping `NUM_BYTES_PER_SHA256 = 32` caused a lot of issues
in the background. Changing it to 31 to match Noir does mean slightly
different encoding, but many fewer changes across the repo (and
hopefully less confusion).~
As of 21.3, due to changes in master, it's now cleaner to keep
`NUM_BYTES_PER_SHA256 = 32` and be sure to truncate and pad all SHA
hashes which touch the Noir circuits.
Since I've kept the hash output as an array of one in Noir, there are
many tuples of one in ts (for the above reasoning) - this can be changed
if preferable.

Methods:

- `toTruncField`: Mirrors Noir's `field_from_bytes_32_trunc` to convert
to a field element - used in place of old method `to2Fields` (tested in
`free_funcs.test.ts`).
- `fromTruncField`: Converts the above back to a 31 byte buffer (tested
as above).

---

ledwards2225 committed:
feat: simplified bb Honk interface (#5319)

Purpose of this PR is to clarify and simplify the bb interface for
constructing and verifying Honk proofs (both UltraHonk and
GoblinUltraHonk).

A similar flow was previously achieved somewhat indirectly through the
`goblin` class via `goblin.accumulate`. This was simply done for
convenience a while back and is not the right thing long term. The new
Honk flows are simplified and do not make use of anything like the
`AcirComposer` used for Plonk.

I have only added flows of the prove-AND-verify variety, i.e. more logic
will be needed in order to separate out the proving and verifying (a la
the prove-THEN-verify flows for Plonk). This includes serialization of
proving and verification keys.

nventuro commented:
I'd strongly recommend holding this off until we can test this from noir directly. There we'd simply mock the oracle response and test the noir return value from noir itself. The end to end approach has a million in-between layers that make the test a) way harder to write b) way harder to read and ensure correctness c) way slower.

Commit:
feat: ZeroMorph working with IPA and integration with ECCVM (#5246)

Resolves https://github.com/AztecProtocol/barretenberg/issues/769.
Resolves https://github.com/AztecProtocol/barretenberg/issues/782.

Generalise Zeromorph further to be able to instantiate with both KZG and
IPA and switch ECCVM to use Zeromorph + IPA. This PR also fixes a small
inconsistency in Zeromorph where we assumed that first element of the
SRS is always going to be [1]_1 which is not the case if we work on
Grumpkin. Unskip and complete eccvm transcript tests.

spalladino committed:
chore: Remove unused FunctionLeafPreimage struct (#5354)

No longer used after the new contract deployment flow.

spalladino committed:
fix: Generate noir interface for constructors (#5352)

With the last changes for contract deployment, a contract **can** call
another contract's initializer.

Co-authored-by: ludamad <adam.domurad@gmail.com>

ledwards2225 committed:
chore: Name change: gen perm sort to delta range constraint (#5378)

Change the naming convention "gen perm sort" to "delta range
constraint". The name GenPermSort is misleading. The relation actually
is checking that the difference (delta) between wire values is no more
than 3. The original name stems from the context in which this relation
is used but says nothing about the actual constraint being applied.

Closes https://github.com/AztecProtocol/barretenberg/issues/919

Commit:
git subrepo push --branch=master barretenberg

subrepo:
  subdir:   "barretenberg"
  merged:   "fd0490da1"
upstream:
  origin:   "https://github.com/AztecProtocol/barretenberg"
  branch:   "master"
  commit:   "fd0490da1"
git-subrepo:
  version:  "0.4.6"
  origin:   "???"
  commit:   "???" [skip ci]

Commit:
chore: replace relative paths to noir-protocol-circuits

Commit:
git_subrepo.sh: Fix parent in .gitrepo file. [skip ci]

Commit:
git subrepo push --branch=master noir-projects/aztec-nr

subrepo:
  subdir:   "noir-projects/aztec-nr"
  merged:   "423ea499e"
upstream:
  origin:   "https://github.com/AztecProtocol/aztec-nr"
  branch:   "master"
  commit:   "423ea499e"
git-subrepo:
  version:  "0.4.6"
  origin:   "???"
  commit:   "???" [skip ci]

sklppy88 committed:
refactor: make get_notes fail if returning no notes #4988 (#5320)

Resolves #4988.

benesjan commented on yarn-project/end-to-end/src/e2e_note_getter.test.ts:
```diff
@@ -16,11 +16,12 @@ function unwrapOptions<T>(options: NoirOption<T>[]): T[] {
 }
 
 describe('e2e_note_getter', () => {
```
To make the tests run in CI you have to add the test in `/.circleci/config.yml`

benesjan replied:
I am retarded and somehow thought this is a completely new test. Ignore me

LHerskind committed:
feat(AuthWit): chain_id and version in hash (#5331)

Fixes #5074 by including the `chain_id` and `version` as part of the message hash that we are signing over in the authwits.

LHerskind committed:
chore: skip slither in docker (#5384)

Skipping slither runs in the docker for less pain with the changes happening all the time.

PhilWindle committed:
feat: Dynamic proving (#5346)

This PR introduces a new proving orhestration system enabling a block to
be proven concurrently with simulation. It also creates the
`prover-client` package and moves some of the proving functionality to
this package from `sequencer-client`.

benesjan commented on noir-projects/aztec-nr/aztec/src/note/note_getter.nr:
```diff
@@ -121,6 +123,13 @@ pub fn get_notes<Note, N, FILTER_ARGS>(
+            if (write_index != i) {
+                opt_notes[write_index] = Option::some(note);
+                opt_notes[i] = Option::none();
+            }
```
Could you document here what this is doing? I would be very confused what's going on if I didn't have the context.

Commit:
Add comments

Commit:
tweak

benesjan committed:
refactor: messaging naming fixes (#5383)

Some naming fixes we discussed in our team chat and which Lasse
[disliked](https://github.com/AztecProtocol/aztec-packages/pull/5206#discussion_r1528245150)
in my purge PR a few days ago.
\+ nuked some stale stuff from contracts.

Commit:
fix

sklppy88 committed:
Merge branch 'master' into ek/refactor/get-notes-returns-non-sparse-array

sklppy88 committed:
Update note_getter.nr

sklppy88 committed:
Merge branch 'master' into ek/refactor/get-notes-returns-non-sparse-array

sklppy88 committed:
Merge branch 'master' into ek/refactor/get-notes-returns-non-sparse-array

sklppy88 committed:
Merge branch 'master' into ek/refactor/get-notes-returns-non-sparse-array

benesjan commented on noir-projects/noir-contracts/contracts/test_contract/src/main.nr:
```diff
@@ -130,6 +130,44 @@ contract Test {
         emit_unencrypted_log_from_private(&mut context, opt_notes[1].unwrap().value);
     }
 
+    pub fn filter_that_creates_sparse_array(
```
Could you create an issue with the recommendation from @nventuro and point here and above create_and_get_many_notes_with_filter to nuke this once we have proper noir testing?

benesjan commented on yarn-project/end-to-end/src/e2e_note_getter.test.ts:
```diff
@@ -252,6 +253,28 @@ describe('e2e_note_getter', () => {
         expect(viewNotesManyResult.sort()).toEqual([BigInt(VALUE), BigInt(VALUE + 1)]);
       }, 45_000);
+
+      it('get_notes should collapse sparse arrays', async () => {
```
Would also link the issue here.

benesjan reviewed (approved):
Looks good to me and I would merge this to not having it hanging around. Just please link the issue and assign it to our team so that we don't forget replacing the e2e test with a unit test.

sklppy88 committed:
Merge branch 'master' into ek/refactor/get-notes-returns-non-sparse-array

Commit:
fix

fix

fix

Commit:
cleanup

Commit:
cleanup

sklppy88 committed:
Merge branch 'master' into ek/refactor/get-notes-returns-non-sparse-array

benesjan commented on noir-projects/aztec-nr/aztec/src/note/note_getter.nr:
```diff
@@ -2,13 +2,14 @@ use dep::protocol_types::{
 };
 use crate::context::PrivateContext;
 use crate::note::{
+    note_header::NoteHeader,
```
I don't see NoteHeader used anywhere in the file diff.

Commit:
addressing comments

benesjan commented on noir-projects/aztec-nr/aztec/src/note/note_getter.nr:
```diff
@@ -1,14 +1,14 @@
     note_interface::NoteInterface, note_viewer_options::NoteViewerOptions,
-    utils::compute_note_hash_for_consumption
+    utils::compute_note_hash_for_consumption,
 };
```
Can you run formatter? This changes seem to be purely formatting related

benesjan commented on noir-projects/aztec-nr/tests/src/note_getter_test.nr:
```diff
@@ -0,0 +1,124 @@
+    let mut test_note = TestNote::new(1337);
+    test_note.header = NoteHeader::new(AztecAddress::from_field(69), 0, 42);
+
+    let mut opt_notes: [Option<TestNote>; MAX_NOTE_HASH_READ_REQUESTS_PER_CALL] = [Option::none(); MAX_NOTE_HASH_READ_REQUESTS_PER_CALL];
```
```suggestion
    let mut opt_notes = [Option::none(); MAX_NOTE_HASH_READ_REQUESTS_PER_CALL];
```

Just a heads up. I think you don't need to specify the type here and Noir should derive it on its own.

sklppy88 committed:
Update noir-projects/aztec-nr/tests/src/note_getter_test.nr

Co-authored-by: Jan Beneš <janbenes1234@gmail.com>

Commit:
fmt

sklppy88 committed:
Merge branch 'master' into ek/refactor/get-notes-returns-non-sparse-array

benesjan commented on noir-projects/aztec-nr/tests/src/mock/test_note.nr:
```diff
@@ -0,0 +1,63 @@
+    note_getter_options::{NoteGetterOptions, Select, Sort, SortOrder, Comparator, NoteStatus, PropertySelector},
+    note_interface::NoteInterface, note_viewer_options::NoteViewerOptions,
+    utils::compute_note_hash_for_consumption, note_getter::_get_notes_constrain_get_notes_internal
+};
```
A lot of the imports here seem to be unused.

benesjan commented on noir-projects/aztec-nr/tests/src/note_getter_test.nr:
```diff
@@ -0,0 +1,131 @@
+
+    let storage_slot: Field = 42;
+
+    let mut opt_notes: [Option<TestNote>; MAX_NOTE_HASH_READ_REQUESTS_PER_CALL] = [Option::none(); MAX_NOTE_HASH_READ_REQUESTS_PER_CALL];
```
This array is declared twice for seemingly no reason.

benesjan reviewed (changes requested):
Looks good. Just please clean up the 2 things.

Commit:
address comments

Commit:
fix

sklppy88 committed:
Merge branch 'master' into ek/refactor/get-notes-returns-non-sparse-array

sklppy88 committed:
Merge branch 'master' into ek/refactor/get-notes-returns-non-sparse-array

Commit:
why
